mantidproject/mantid
# Add calibration file input to ISIS Reflectometry GUI

**Description of work.**
This PR adds a new `CalibrationFile` input to the Instrument Settings tab of the Reflectometry interface. This input is blank by default, but when a file path is provided a calibration step will automatically be applied for normal reductions, live reductions or when loading preview data.

If the Debug checkbox on the Experiment Settings tab is ticked then the calibration step will output a table workspace containing the calibration data that is being applied. If the Debug checkbox is unticked then the calibration table workspace is removed from the ADS before the calibration step completes.

Although the calibration file path is optional data, I didn't use a `boost::optional` for holding the data in the Instrument model as it seemed sufficient to hold it as an empty string.

**Report to:** Max at ISIS

**To test:**

1) Open the ISIS Reflectometry interface and set the instrument to INTER.
2) On the Experiment Settings tab, tick the Debug checkbox. This will make sure that a calibration table is output to the ADS when the calibration step is run. Also set the AnalysisMode to `MultiDetectorAnalysis`.
3) Download the example calibration data from issue #32022. On the Instrument Settings tab, find the CalibrationFile input at the bottom of the tab and click the button to browse to the calibration file provided in the example data (`pixel_centers_repeat.dat`).
4) Go to the Runs tab. In the Runs table (on the right hand side) enter run `61135` and angle `0.5` into the first row of the table. Select the row and click to process. The reduction should complete successfully and the calibration step should have been applied (see messages in the panel, output calibration workspace and also `ReflectometryISISCalibration` should be in the history of workspace `IvsQ_binned_61135`).
5) Clear all the workspaces and go to the Reduction Preview tab. Load the `INTER45455_inst` dataset and again check that the calibration step is applied (there should be messages in the panel plus an output calibration workspace).
6) Finally, check live data reduction by clicking the Start Monitor button on the Runs tab (this will need to be tested at ISIS while instruments are live). The calibration step may throw errors if the test calibration file doesn't contain the right entries for the experiment, however that's sufficient to prove that calibration is being applied to live data reductions.
7) Test running reductions with the Debug checkbox unticked, which should prevent the calibration table workspace being left in the ADS. Also check that calibration is not applied when the CalibrationFile input is left blank.

Fixes #34111.





---

#### Reviewer ####

Please comment on the following ([full description](http://developer.mantidproject.org/ReviewingAPullRequest.html)):

##### Code Review #####

- Is the code of an acceptable quality?
- Does the code conform to the [coding standards](http://developer.mantidproject.org/Standards/)?
- Are the unit tests small and test the class in isolation?
- If there is GUI work does it follow the [GUI standards](http://developer.mantidproject.org/Standards/GUIStandards.html)?
- If there are changes in the release notes then do they describe the changes appropriately?
- Are the release notes saved in a separate file, using Issue or PR number for file name and in the correct location?

##### Functional Tests #####

- Do changes function as described? Add comments below that describe the tests performed?
- Do the changes handle unexpected situations, e.g. bad input?
- Has the relevant (user and developer) documentation been added/updated?

Does everything look good? Mark the review as **Approve**. A member of `@mantidproject/gatekeepers` will take care of it.
